src/rate_limiter.rs

```diff
@@ -0,0 +1,224 @@
+        // others.
+        let leaking_client_count = self.used.len();
+        let mut quota = leaked_units / leaking_client_count as u64;
+        let mut entries: Vec<(u64, IpAddr)> = self.used
+            .iter()
+            .map(|(ip_addr, used)| (*used, *ip_addr))
+            .collect();
+        entries.sort();
+        self.used.clear();
+        for (index, (used, client)) in entries.into_iter().enumerate() {
+            if used < quota {
+                leaked_units -= used;
+                // The divisor will never be `0` as such a case would need all entries to be below
+                // their quota (i.e. all usage has fully leaked) and would have results in an early
+                // return in the above sum check already.
+                quota = leaked_units / (leaking_client_count - index - 1) as u64;
+            } else {
+                let _ = self.used.insert(client, used - quota);
+            }
+        }
+    }
+}
+
+impl Default for RateLimiter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(all(test, feature = "use-mock-crust"))]
+mod tests {
+    use super::*;
+    use fake_clock::FakeClock;
+    use messages::Request;
+    use rand;
+    use types::MessageId;
+
+    #[test]
+    fn add_message() {
+        // First client fills the `RateLimiter` with get requests.
+        let mut rate_limiter = RateLimiter::new();
+        let client_1 = IpAddr::from([0, 0, 0, 0]);
+        let get_req_payload = unwrap!(serialisation::serialise(
+            &UserMessage::Request(Request::GetIData {
+                name: rand::random(),
+                msg_id: MessageId::new(),
+        })));
+        let fill_full_iterations = CAPACITY / CLIENT_GET_CHARGE;
+        for _ in 0..fill_full_iterations {
+            unwrap!(rate_limiter.add_message(1, &client_1, 1, 0, &get_req_payload));
+        }
+
+        // Check a second client can't add a message just now.
+        let client_2 = IpAddr::from([1, 1, 1, 1]);
+        match rate_limiter.add_message(1, &client_2, 1, 0, &get_req_payload) {
+            Err(RoutingError::ExceedsRateLimit) => {}
+            _ => panic!("unexpected result"),
+        }
+
+        // Wait until enough has drained to allow the second client's request to succeed.
+        let wait_millis = CLIENT_GET_CHARGE * 1000 / RATE as u64;
+        FakeClock::advance_time(wait_millis);
+        unwrap!(rate_limiter.add_message(10, &client_2, 1, 0, &get_req_payload));
+
+        // Wait for the same period, but now try adding invalid messages.
+        FakeClock::advance_time(wait_millis);
+        let all_zero_paylod = vec![0u8; CLIENT_GET_CHARGE as usize];
+        match rate_limiter.add_message(10, &client_2, 2, 0, &all_zero_paylod) {
+            Err(RoutingError::InvalidMessage) => {}
+            _ => panic!("unexpected result"),
+        }
+
+        unwrap!(rate_limiter.add_message(2, &client_2, 1, 0, &get_req_payload));
```

Comment:
better have a
```rust
        match rate_limiter.add_message(10, &client_2, 1, 0, &get_req_payload) {
            Err(RoutingError::ExceedsRateLimit) => {}
            _ => panic!("unexpected result"),
        }
```
test first to confirm the per-client usage check has been carried out properly